src/main/java/com/google/devtools/build/lib/remote/LoggingInterceptor.java

```diff
@@ -0,0 +1,107 @@
+import javax.annotation.Nullable;
+
+/** Client interceptor for logging details of certain gRPC calls. */
+public class LoggingInterceptor implements ClientInterceptor {
+
+  /**
+   * Returns a {@link LoggingHandler} to handle logging details for the specific method.
+   *
+   * @param method Method to return handler for.
+   * @return A handler for the given method. Returns null if the method has no specific handler.
+   */
+  protected @Nullable LoggingHandler selectHandler(MethodDescriptor method) {
+    // TODO(cdlee): add handlers for methods
+    return null;
+  }
+
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+      MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+    LoggingHandler<ReqT, RespT> handler = selectHandler(method);
+    if (handler != null) {
+      return new LoggingForwardingCall<>(next.newCall(method, callOptions), handler, method);
+    } else {
+      return next.newCall(method, callOptions);
+    }
+  }
+
+  /**
+   * Wraps client call to log call details by building a {@link LogEntry} and writing it to a log.
+   */
+  private static class LoggingForwardingCall<ReqT, RespT>
+      extends ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT> {
+    private LoggingHandler<ReqT, RespT> handler;
+    private LogEntry.Builder entryBuilder = LogEntry.newBuilder();
+
+    protected LoggingForwardingCall(
+        ClientCall<ReqT, RespT> delegate,
+        LoggingHandler<ReqT, RespT> handler,
+        MethodDescriptor<ReqT, RespT> method) {
+      super(delegate);
+      this.handler = handler;
+      entryBuilder.setMethodName(method.getFullMethodName());
+    }
+
+    @Override
+    public void start(Listener<RespT> responseListener, Metadata headers) {
+      try {
+        entryBuilder.setMetadata(extractRequestMetadata(headers));
+      } catch (IllegalStateException e) {
+        // Don't set RequestMetadata field if it is not present.
+      }
+      super.start(
+          new ForwardingClientCallListener.SimpleForwardingClientCallListener<RespT>(
+              responseListener) {
+            @Override
+            public void onMessage(RespT message) {
+              handler.handleResp(message);
+              super.onMessage(message);
+            }
+
+            @Override
+            public void onClose(Status status, Metadata trailers) {
+              entryBuilder.setStatus(makeStatusProto(status));
+              LogEntry entry = entryBuilder.mergeFrom(handler.getEntry()).build();
+              // TODO(cdlee): Actually log the entry.
+              super.onClose(status, trailers);
+            }
+          },
+          headers);
+    }
+
+    @Override
+    public void sendMessage(ReqT message) {
+      handler.handleReq(message);
+      super.sendMessage(message);
+    }
+  }
+
+  // Converts grpc.io.Status to com.google.rpc.Status for logging.
+  private static com.google.rpc.Status makeStatusProto(Status status) {
```

Comment:
You can use this https://cs.corp.google.com/piper///depot/google3/third_party/java_src/grpc/protobuf/src/main/java/io/grpc/protobuf/StatusProto.java?rcl=179223655&l=145, it's open source.

Reply:
Unfortunately, I don't think this fits my need because it is private and it extracts the status proto from trailers rather than from the status object. Also, from what I can tell (not really too familiar with gRPC), this trailer isn't automatically populated, nor do many people set it.